rust/kernel/delay.rs

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Delay functions for operations like sleeping.
+//!
+//! C header: [`include/linux/delay.h`](../../../../include/linux/delay.h)
+
+use crate::bindings;
+use core::{cmp::min, time::Duration};
+
+/// Sleeps safely even with waitqueue interruptions.
+///
+/// This function forwards the call to the C side `msleep` function. As a result,
+/// `duration` will be rounded up to the nearest millisecond if granularity less
+/// than a millisecond is provided. Any [`Duration`] that exceeds
+/// [`c_uint::MAX`][core::ffi::c_uint::MAX] in milliseconds is saturated.
+pub fn coarse_sleep(duration: Duration) {
+    let duration_as_nanos = duration.as_nanos();
+    let milli_as_nanos = Duration::MILLISECOND.as_nanos();
+    let round = if duration_as_nanos % milli_as_nanos > 0 {
+        1
+    } else {
+        0
+    };
```

Comment:
Is there a problem with `duration_as_nanos.saturating_add(milli_as_nanos - 1) / milli_as_nanos`?

Reply:
Okay I've made the change!